### PR TITLE
improve handling of glued object extraction

### PR DESCRIPTION
**Pull request by MichaelChirico**
1. Pull declaration of the `.transformer` function outside the `for (cl in glue_calls)` loop
2. Use the `.envir` environment to remove the dependency of the `.transformer` on globals inside `extract_glued_symbols`
3. Operate directly on the `getParseData()` output for this symbol case rather than converting to XML and then searching an XPath
4. Use `names(env)` over `ls(env, all.names=TRUE)` which I found is somehow [20x faster](https://twitter.com/michael_chirico/status/1577129104515530753)

I tried testing the performance on `dplyr`, but it's mostly using `glue()` unqualified, so I haven't had any luck yet. We might try loosening that restriction.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
Benchmark-ish. Did a find-and-replace in one dplyr file for `glue(` --> `glue::glue(` to ensure `object_usage_linter()` is actually evaluating `glue()` calls. Then ran `lint()` 40 times:

```r
# on main
system.time(replicate(40, simplify=FALSE, lint("dplyr/R/join-by.R", object_usage_linter())))
#     user  system elapsed
#   63.096   0.331  63.701

# on this branch
system.time(replicate(40, simplify=FALSE, lint("dplyr/R/join-by.R", object_usage_linter())))
#     user  system elapsed
#   59.484   0.204  59.891
```

So a small speed-up (5-6%) in typical usages. It should be a lot better in pathological cases (especially a `{...}` expression with a lot of variables)

File: R/object_usage_linter.R [hunk 1]
```diff
@@ -196,7 +182,28 @@ extract_glued_symbols <- function(expr) {
...
+    warning = function(...) NULL
+  )
+  if (is.null(parsed_text)) {
+    return("")
+  }
+  parse_data <- utils::getParseData(parsed_text)
+  # covers NULL & NA cases
+  if (nrow(parse_data) == 0L) {
+    return("")
+  }
```
Comment (IndrajeetPatil) [inline R/object_usage_linter.R:201]: Are there existing tests indirectly covering the two conditionals here?

File: R/object_usage_linter.R [hunk 2]
```diff
@@ -196,7 +182,28 @@ extract_glued_symbols <- function(expr) {
...
+  symbols <- parse_data$text[parse_data$token == "SYMBOL"]
```
Comment (AshesITR) [inline R/object_usage_linter.R:202]: Does this approach differentiate between `x` and `` `x` ``?

Comment (MichaelChirico) [reply R/object_usage_linter.R:202]: Hmm I think it doesn't, but neither did the old approach.

**Review by AshesITR — APPROVED** [review]

**Pull request merged**